### ISSUE TITLE
Compiler exception: "Too many nodes written to output"

### ISSUE THREAD

**Issue by Samotorcan** [Microsoft/TypeScript#13664]
**TypeScript Version:**  2.1.5

**Code**

```ts
declare let doSomething;

for (let a1 of [])
    for (let a2 of a1.someArray)
        doSomething(() => a2);
```

tsconfig looks like this:
```
{
    "compilerOptions": {
        "moduleResolution": "node",
        "target": "es5"
    },
    "files": [
        "test.ts"
    ]
}
```

**Expected behavior:**
Should compile.

**Actual behavior:**
Typescript compiler fails on an assertion, throwing the following exception:
```
Error: Debug Failure. False expression: Too many nodes written to output.
    at Object.assert (...\built\local\tsc.js:2707:23)
    at extractSingleNode (...\built\local\tsc.js:38414:15)
    at Object.visitNode (...\built\local\tsc.js:38091:54)
    at convertForOfToFor (...\built\local\tsc.js:42367:36)
    at convertIterationStatementBodyIfNecessary (...\built\local\tsc.js:42440:40)
    at visitForOfStatement (...\built\local\tsc.js:42320:20)
    at visitJavaScript (...\built\local\tsc.js:41590:28)
    at visitorWorker (...\built\local\tsc.js:41518:24)
    at dispatcher (...\built\local\tsc.js:41439:19)
    at saveStateAndInvoke (...\built\local\tsc.js:41457:27)
```

The problem goes away if the first loop is written with curly brackets.
```ts
declare let doSomething;

for (let a1 of []) {
    for (let a2 of a1.someArray)
        doSomething(() => a2);
}
```

**Comment by RyanCavanaugh** [comment]
Same title, repro, and callstack as #12961 - I think we can call this a duplicate

**Comment by Samotorcan** [comment]
Just to clarify, #12961 was fixed and closed in version 2.1.5 while this example fails to compile and throws and error in version 2.1.5.

**Comment by vladima** [comment]
I can repro this on the latest master.
